### PR TITLE
Fix incorrect dedup of non-deterministic functions

### PR DESCRIPTION
**Pull request by martint**
Fixes https://github.com/trinodb/trino/issues/22683

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# General
* Fix incorrect evaluation of repeated non-deterministic functions. ({issue}`22683`)
```

### CODE REVIEW

**Review by hashhar — APPROVED** [review]

**Comment by martint** [comment]
@hashhar, I made a few additional changes. Mind taking a look again?

File: core/trino-main/src/main/java/io/trino/operator/scalar/ArrayReduceFunction.java [hunk 1]
```diff
@@ -60,7 +60,6 @@ private ArrayReduceFunction()
                         .build())
                 .nullable()
                 .argumentNullability(false, true, false, false)
-                .nondeterministic()
```
Comment (hashhar) [inline core/trino-main/src/main/java/io/trino/operator/scalar/ArrayReduceFunction.java:63]: > Non-determinism depends on the body of the lambda expressions
passed to the function.
> That's handled elsewhere.

Where can I find that?

**Review by hashhar — APPROVED** [review]
I don't know/cant verify impact of " Mark array and map processing functions as deterministic " since I couldn't find the code which determines a lambda as deterministic or not.

LGTM otherwise.

File: core/trino-main/src/main/java/io/trino/operator/scalar/ArrayReduceFunction.java (hunk 1 above)
Comment (martint) [reply core/trino-main/src/main/java/io/trino/operator/scalar/ArrayReduceFunction.java:63]: It’s handled by DeterminismEvaluator.
It’s may not be immediately obvious that lambdas are handled, but that class walks the expression tree and returns “not deterministic” if any call contained in the expression is non-deterministic.

Comment (hashhar) [reply core/trino-main/src/main/java/io/trino/operator/scalar/ArrayReduceFunction.java:63]: yup, I see it now. It's a surprisingly simple implementation too (of the DeterminismEvaluator(s)).
Looks good to merge IMO.

**Review by hashhar — APPROVED** [review]

**Pull request merged**